### ISSUE TITLE
web workflow broken on ESP32 V2 with 8.0.2

### ISSUE THREAD

**Issue by vladak** [adafruit/circuitpython#7636]
### CircuitPython version

```python
CircuitPython 8.0.2 on Adafruit Feather ESP32 V2
```

### Code/REPL

```python
N/A
```

### Behavior

N/A

### Description

In CircuitPython 8.0.2 with ESP32 V2, whenever a file is uploaded (PUT) or deleted (DELETE) the web workflow gets broken because the board no longer listens for TCP connections on port 80.

Downgrading to 8.0.0 beta 4 fixes the problem. https://github.com/adafruit/circuitpython/commit/f8ac1d9261c087d1f65886036289c2121c32cc0f#diff-43f36ea6fd6b4d28f257e55f3f1f29a9890e72c88d7b56164d77ab5691b704cc looks suspicious to me.

### Additional information

See https://forums.adafruit.com/viewtopic.php?p=961741 for more details.

**Comment by dhalbert** [comment]
@tannewt Would you mind taking a look at this? Thanks.

**Comment by tannewt** [comment]
@vladak What versions did you try? 8.0.0? 8.0.1?

A put and delete shouldn't change the reset reason. So I don't think it is the cause of the issue you are seeing.

**Comment by RetiredWizard** [comment]
I've been doing a little testing on the Huzzah32 (thanks for the kickstart @dhalbert :blush:). I didn't reproduce exactly what it sounds like @vladak is describing but I did find something that maybe they are running into.

On the Huzzah32 with the default code.py file when the board boots it sits at the Press any key to enter the REPL prompt. Web workflow works fine and you can access the serial terminal. However if you are at that initial "Press any key" prompt and you use the web workflow file browser to upload a file, the board does an auto-reload and ends up back at the "Press any key" prompt. Now, the web workflow serial terminal is no longer accessible. The rest of the web workflow seems to work find, it's just the serial terminal (and full code editor) that appear to have trouble until the board is reset.

**Comment by RetiredWizard** [comment]
The serial terminal behavior started with rc.0, it doesn't happen on beta.6.

**Comment by vladak** [comment]
> I could not reproduce this on a Feather ESP32 V2 or a Metro ESP32-S2. I uploaded files and deleted files, and did not see a problem. Is your code.py doing some networky thing? Mine is not.

`code.py` contains just this:
```
print("Hello World!")
```

**Comment by dhalbert** [comment]
Could you erase the flash completely and load 8.0.2 fresh? I know this will be a bit of a nuisance. What browser are you using (I don't think it should matter, but...)

**Comment by vladak** [comment]
Here's what I have tried so far:

Version | Problem present
--- | ---
8.0.2 | Yes
8.0.0 | Yes
8.0.0-rc0 | Yes
8.0.0-beta6 | No
8.0.0-beta5 |No
8.0.0-beta4 |No

so it seems to be introduced in 8.0.0-rc0.

**Comment by dhalbert** [comment]
What files were you uploading and deleting? I was uploading and deleting datafiles, not code.py or some other used .py file.

**Comment by vladak** [comment]
> Could you erase the flash completely and load 8.0.2 fresh? I know this will be a bit of a nuisance. What browser are you using (I don't think it should matter, but...)

I use REST APIs via `circup` (or rather [my PR](https://github.com/adafruit/circup/pull/163) that introduces web workflow support), however it fails with browser as well (Firefox in my case).

**Comment by vladak** [comment]
After doing
```
esptool.py --port /dev/ttyACM0 erase_flash
esptool.py --port /dev/ttyACM0 write_flash -z 0x0 \
    ~/Downloads/adafruit-circuitpython-adafruit_feather_esp32_v2-en_US-8.0.2.bin
```
I created the `settings.toml` file using the guide on https://learn.adafruit.com/circuitpython-with-esp32-quick-start/setting-up-web-workflow , verified it is reachable via ping and TCP on port 80 and retried. It fails in the same way as before.

**Comment by vladak** [comment]
> What files were you uploading and deleting? I was uploading and deleting datafiles, not code.py or some other used .py file.

I am uploading Python libraries.

**Comment by dhalbert** [comment]
sorry for the continuing q's, still trying to reporduce. Are you addressing the board with a DNS name, or a numeric IP? I have trouble with cached entries in my browser, so I always use a numeric IP.

**Comment by vladak** [comment]
Using numeric IP address.

**Comment by vladak** [comment]
Happens also when `code.py` is changed with:
```python
import requests
from requests.auth import HTTPBasicAuth


auth = HTTPBasicAuth("", "XXX")
r = requests.put("http://172.40.0.10/fs/code.py",
                 'print("Hello ESP32 V2!")',
                 auth=auth)
```

**Comment by RetiredWizard** [comment]
For the serial terminal issue I was seeing, I tracked it down to https://github.com/adafruit/circuitpython/commit/40534da92a8e385b9a04b8d0ff14f8365b2e1f5a and in particular this change:
```
  uint64_t start_ticks = supervisor_ticks_ms64();

     // Allow timeouts and interrupts
-    while (newsoc == -1 && !timed_out) {
+    while (newsoc == -1 &&
+           !timed_out &&
+          !mp_hal_is_interrupted()) {
         if (self->timeout_ms != (uint)-1 && self->timeout_ms != 0) {
             timed_out = supervisor_ticks_ms64() - start_ticks >= self->timeout_ms;
         }
```

@vladak , This probably isn't the fix but I'd be curious if the attached adafruit_feather_esp32_v2 firmware resolves the issues you are seeing.

[firmware.zip](https://github.com/adafruit/circuitpython/files/10819131/firmware.zip)

**Comment by RetiredWizard** [comment]
On a somewhat unrelated note, when I was looking at the commit in question it looked to me like the this change:
```
-            write(socket_change_fd, &fd, sizeof(fd));
+            // Write must be 8 bytes for an eventfd.
+            uint64_t signal = 1;
+            write(socket_change_fd, &signal, sizeof(signal));
  ```
should have set signal to fd, but I don't have any clue as to what the code is actually doing here....

**Comment by dhalbert** [comment]
> should have set signal to fd

The value of signal doesn't seem to be used anywhere, and the size of the write was wrong earlier (it was 4 bytes instead of 8, which is required). Maybe fixing that broke something this other thing.

The first fix, adding `!mp_hal_is_interrupted()`, allows ctrl-C, I don't think that would affect this. Is that the change you made in the attached firmware.zip?

**Comment by RetiredWizard** [comment]
Yep, I removed the !mp_hal_is_intrrupted() and my issue went away.

**Comment by dhalbert** [comment]
> Yep, I removed the !mp_hal_is_intrrupted() and my issue went away.

That is weird, it would normally be always true (ctrl-c not typed).

**Comment by tannewt** [comment]
Autoreload could cause interrupt too I think. I think there needs to be a check after that loop to ensure that newsoc is valid. If not, return EAGAIN like the check inside the loop does.

**Comment by tannewt** [comment]
Or, looking more closely, `lwip_fcntl` should happen later after we know `newsoc` is valid.

**Comment by Neradoc** [comment]
I get the issue on a FunHouse too, when I try to upload a file, it fails as it should because the drive is mounted, but **still auto-reloads**, and after that the web workflow stops responding.
The issue also happens when creating a directory but ONLY if the directory is really created, not if it already exists, or if the drive is mounted. Which is also the only case where it triggers an auto-reload.

**Comment by vladak** [comment]
> @vladak , This probably isn't the fix but I'd be curious if the attached adafruit_feather_esp32_v2 firmware resolves the issues you are seeing.
> 
> [firmware.zip](https://github.com/adafruit/circuitpython/files/10819131/firmware.zip)

The code that identifies as `8.1.0-alpha.2-17-g00a03c323-dirty` indeed resolves the issue. Using my version of `circup` that supports web workflow I can install and uninstall libraries consisting of multiple files without issue. The `code.py` uploader above (which I modified a bit to run the upload in a short cycle) works as well. Hurray !

**Comment by dhalbert** [comment]
@vladak, here is a fix similar to @RetiredWizard's, and in addition it moves `lwip_fcntl()` to a place where `newsoc` is know to be good. Would you mind testing this? Thanks very much. If it works I will PR this to 8.0.x.

[feather-esp32v2.bin.zip](https://github.com/adafruit/circuitpython/files/10886070/feather-esp32v2.bin.zip)

**Comment by Neradoc** [comment]
> @vladak, here is a fix similar to @RetiredWizard's, and in addition it moves `lwip_fcntl()` to a place where `newsoc` is know to be good. Would you mind testing this? Thanks very much. If it works I will PR this to 8.0.x.
> 
> [feather-esp32v2.bin.zip](https://github.com/adafruit/circuitpython/files/10886070/feather-esp32v2.bin.zip)

That seems to work, the same tests I ran before do not trigger the issue with this build.
I also tested the code from the circup web workflow PR.

**Comment by vladak** [comment]
> @vladak, here is a fix similar to @RetiredWizard's, and in addition it moves `lwip_fcntl()` to a place where `newsoc` is know to be good. Would you mind testing this? Thanks very much. If it works I will PR this to 8.0.x.
> 
> [feather-esp32v2.bin.zip](https://github.com/adafruit/circuitpython/files/10886070/feather-esp32v2.bin.zip)

The firmware reports itself as `CircuitPython 8.0.2-2-gc7f485d7d-dirty on 2023-03-03` and seems to be working fine w.r.t. multiple PUT/DELETE requests.

**Comment by tannewt** [comment]
Fixed in #7679
